packages/medusa/src/services/batch-job.ts

```diff
@@ -71,6 +136,243 @@ class BatchJobService extends TransactionBaseService<BatchJobService> {
       }
     )
   }
+
+  async validateBatchContext(
+    type: string,
+    context: Record<string, unknown>
+  ): Promise<Record<string, unknown>> {
+    // TODO validate context with batch job strategy corresponding to the given type
+
+    return context
+  }
+
+  async create(data: BatchJobCreateProps): Promise<BatchJob> {
+    return await this.atomicPhase_(async (manager) => {
+      const batchJobRepo: BatchJobRepository = manager.getCustomRepository(
+        this.batchJobRepository_
+      )
+
+      const validatedContext = await this.validateBatchContext(
+        data.type,
+        data.context
+      )
+
+      const toCreate = {
+        status: BatchJobStatus.CREATED,
+        ...data,
+        context: validatedContext,
+      } as DeepPartial<BatchJob>
+
+      const batchJob = await batchJobRepo.create(toCreate)
+      const result = await batchJobRepo.save(batchJob)
+
+      await this.eventBus_
+        .withTransaction(manager)
+        .emit(BatchJobService.Events.CREATED, {
+          id: result.id,
+        })
+
+      return result
+    })
+  }
+
+  async update(
+    batchJobId: string,
+    data: BatchJobUpdateProps
+  ): Promise<BatchJob> {
+    return await this.atomicPhase_(async (manager) => {
+      const batchJobRepo: BatchJobRepository = manager.getCustomRepository(
+        this.batchJobRepository_
+      )
+
+      let batchJob = await this.retrieve(batchJobId)
+
+      const { context, ...rest } = data
+      if (context) {
+        batchJob.context = await this.validateBatchContext(
+          batchJob.type,
+          context
+        )
+      }
+
+      Object.keys(rest)
+        .filter((key) => typeof rest[key] !== `undefined`)
+        .forEach((key) => {
+          batchJob[key] = rest[key]
+        })
+
+      batchJob = await batchJobRepo.save(batchJob)
+
+      await this.eventBus_
+        .withTransaction(manager)
+        .emit(BatchJobService.Events.UPDATED, {
+          id: batchJob.id,
+        })
+
+      return batchJob
+    })
+  }
+
+  protected async updateStatus(
+    batchJobOrId: BatchJob | string,
+    status: BatchJobStatus
```

Comment:
suggestion: name this `nextStatus`?

Reply:
Why not, but status I thought was enough  ^^

Reply:
yeah definitely is very nitty and non-blocking :)